qcore/gsf.py

```diff
@@ -0,0 +1,210 @@
+    nztm_meshgrid = (transformation_matrix @ subdivision_coordinates).T
+    nztm_meshgrid += origin
+    return coordinates.nztm_to_wgs_depth(nztm_meshgrid)
+
+
+def write_fault_to_gsf_file(
+    gsf_file_handle: TextIO,
+    meshgrids: list[np.ndarray],
+    lengths: np.ndarray,
+    widths: np.ndarray,
+    strikes: np.ndarray,
+    dips: np.ndarray,
+    rakes: np.ndarray,
+    resolution: int,
+):
+    """Writes geometry data to a GSF file.
+
+    Parameters
+    ----------
+    gsf_file_handle : TextIO
+        The file handle pointing to the GSF file to write to.
+    meshgrids : list[np.ndarray]
+        List of meshgrid arrays representing fault segments (length: n).
+    lengths : np.ndarray
+        An (n x 1) array of segment lengths.
+    widths : np.ndarray
+        An (n x 1) array of segment widths.
+    strikes : np.ndarray
+        An (n x 1) array of segment strikes.
+    dips : np.ndarray
+        An (n x 1) array of segment dips.
+    rakes : np.ndarray
+        An (n x 1) array of segment rakes.
+    resolution : int
+        Resolution of the meshgrid.
+    """
+    gsf_file_handle.write(
+        "# LON  LAT  DEP(km)  SUB_DX  SUB_DY  LOC_STK  LOC_DIP  LOC_RAKE  SLIP(cm)  INIT_TIME  SEG_NO\n"
+    )
+    number_of_points = sum(meshgrid.shape[0] for meshgrid in meshgrids)
+    gsf_file_handle.write(f"{number_of_points}\n")
+    for i, (length, width, strike, dip, rake, meshgrid) in enumerate(
+        zip(lengths, widths, strikes, dips, rakes, meshgrids)
+    ):
+        strike_step = length / gridpoint_count_in_length(length * 1000, resolution)
+        dip_step = width / gridpoint_count_in_length(width * 1000, resolution)
+        for point in meshgrid:
+            gsf_file_handle.write(
+                f"{point[1]:11.5f} {point[0]:11.5f} {point[2] / 1000:11.5e} {strike_step:11.5e} {dip_step:11.5e} {strike:6.1f} {dip:6.1f} {rake:6.1f} {-1.0:8.2f} {-1.0:8.2f} {i:3d}\n"
+            )
+
+
+def read_gsf(gsf_filepath: str) -> pd.DataFrame:
+    """Parse a GSF file into a pandas DataFrame.
+
+    Parameters
+    ----------
+    gsf_file_handle : TextIO
+        The file handle pointing to the GSF file to read.
+
+    Returns
+    -------
+    pd.DataFrame
+        A DataFrame containing all the points in the GSF file. The DataFrame's columns are
+        - lon (longitude)
+        - lat (latitude)
+        - depth (Kilometres below ground, i.e. depth = -10 indicates a point 10km underground).
+        - sub_dx (The subdivision size in the strike direction)
+        - sub_dy (The subdivision size in the dip direction)
+        - strike
+        - dip
+        - rake
+        - slip (nearly always -1)
+        - init_time (nearly always -1)
+        - seg_no (the fault segment this point belongs to)
+    """
+    with open(gsf_filepath, mode="r", encoding="utf-8") as gsf_file_handle:
+        # we could use pd.read_csv with the skiprows argument, but it's not
+        # as versatile as simply skipping the first n rows with '#'
```

Comment:
I assume thats because n rows is an unknown length which is different per file?

Reply:
Yeah the number of comment lines prior to the number of points line is not defined. I generate GSF files with just one comment line, rob uses 3 lines, both are acceptable.